File: mobile/src/types/friendship.ts
/**
 * Friendship domain types for the mobile client.
 *
 * These mirror the Pydantic schemas from backend/app/schemas/friendship.py.
 */

/** The friendship state between the current user and a target user. */
export type FriendshipUIStatus =
  | "not_friends"
  | "pending_outgoing"
  | "pending_incoming"
  | "friends";

/** Returned by GET /friends/status/{user_id} */
export interface FriendshipStatusResponse {
  status: FriendshipUIStatus;
  friendship_id: string | null;
}

/** Raw friendship record (used for accept/decline/remove) */
export interface FriendshipResponse {
  id: string;
  requester_user_id: string;
  addressee_user_id: string;
  status: "pending" | "accepted" | "declined";
  created_at: string;
  updated_at: string;
}

/** Public user info embedded in friend entries */
export interface FriendUserInfo {
  id: string;
  full_name: string | null;
  profile_image_url: string | null;
}

/** One entry in the accepted friends list (GET /friends) */
export interface FriendEntry {
  friendship_id: string;
  friend: FriendUserInfo;
  since: string;
}

/** Enriched incoming request (GET /friends/requests/incoming) */
export interface IncomingRequestEntry {
  id: string;
  requester: FriendUserInfo;
  status: "pending";
  created_at: string;
}

/** Enriched outgoing request (GET /friends/requests/outgoing) */
export interface OutgoingRequestEntry {
  id: string;
  addressee: FriendUserInfo;
  status: "pending";
  created_at: string;
}

File: backend/app/schemas/friendship.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.models.friendship import FriendshipStatus


class FriendRequestCreate(BaseModel):
    addressee_user_id: uuid.UUID


class FriendshipResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    requester_user_id: uuid.UUID
    addressee_user_id: uuid.UUID
    status: FriendshipStatus
    created_at: datetime
    updated_at: datetime


class FriendUserInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    full_name: str | None
    profile_image_url: str | None


class FriendEntry(BaseModel):
    friendship_id: uuid.UUID
    friend: FriendUserInfo
    since: datetime


# ---------------------------------------------------------------------------
# Enriched request entries — include the other party's user info for the UI
# so it can show name/avatar without a separate fetch.
# ---------------------------------------------------------------------------


class IncomingRequestEntry(BaseModel):
    """Pending incoming request with the requester's public info."""

    id: uuid.UUID
    requester: FriendUserInfo
    status: FriendshipStatus
    created_at: datetime


class OutgoingRequestEntry(BaseModel):
    """Pending outgoing request with the addressee's public info."""

    id: uuid.UUID
    addressee: FriendUserInfo
    status: FriendshipStatus
    created_at: datetime


# ---------------------------------------------------------------------------
# Friendship status for the mobile UI (used by PublicProfileScreen)
# ---------------------------------------------------------------------------

FriendshipUIStatus = Literal["not_friends", "pending_outgoing", "pending_incoming", "friends"]


class FriendshipStatusResponse(BaseModel):
    status: FriendshipUIStatus
    friendship_id: uuid.UUID | None
